Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -22,18 +24,75 @@ namespace Microsoft.Dafny {
   [System.AttributeUsage(System.AttributeTargets.Field)]
   public class FilledInDuringResolutionAttribute : System.Attribute { }
 
-  public interface INode {
+  public abstract class INode {
 
     /// <summary>
     /// These children should be such that they contain information produced by resolution such as inferred types
     /// and resolved references. However, they should not be so transformed that source location from the initial
     /// program is lost. As an example, the pattern matching compilation may deduplicate nodes from the original AST,
     /// losing source location information, so those transformed nodes should not be returned by this property.
     /// </summary>
-    IEnumerable<INode> Children { get; }
+    public abstract IEnumerable<INode> Children { get; }
+
+    public IToken? StartToken { get; set; } = Token.NoToken;
```

Comment:
I would make a single field `Range` that contains both tokens, that way you can't forget to set one of the two and it reduces clutter when checking the members of an `INode`.